src/site/content/en/blog/webassembly-memory-debugging/index.md

```diff
@@ -0,0 +1,607 @@
+  -fsanitize=address \
+  node_modules/libimagequant/libimagequant.a
+```
+
+This will automatically enable pointer safety checks, but we also want to find potential memory
+leaks. Since we're using ImageQuant as a library rather than a program, there is no "exit point" at
+which Emscripten could automatically validate that all memory has been freed.
+
+Instead, for such cases the LeakSanitizer (included in the AddressSanitizer) provides the functions
+[`__lsan_do_leak_check` and
+`__lsan_do_recoverable_leak_check`](https://emscripten.org/docs/debugging/Sanitizers.html#memory-leaks),
+which can be manually invoked whenever we expect all memory to be freed and want to validate that
+assumption. `__lsan_do_leak_check` is meant to be used at the end of a running application, when you
+want to abort the process in case any leaks are detected, while `__lsan_do_recoverable_leak_check`
+is more suitable for library use-cases like ours, when you want to print leaks to the console, but
+keep the application running regardless.
+
+Let's expose that second helper via Embind so that we can call it from JavaScript at any time:
+
+```cpp/0,12/
+#include <sanitizer/lsan_interface.h>
+
+// …
+
+void free_result() {
+  free(result);
+}
+
+EMSCRIPTEN_BINDINGS(my_module) {
+  function("zx_quantize", &zx_quantize);
+  function("version", &version);
+  function("free_result", &free_result);
+  function("doLeakCheck", &__lsan_do_recoverable_leak_check);
+}
+```
+
+And invoke it from the JavaScript side once we're done with the image. Doing this from the
+JavaScript side, rather than the C++ one, helps to ensure that all the scopes have been
+exited and all the temporary C++ objects were freed by the time we run those checks:
+
+```typescript/7/
+  // …
+
+  const result = opts.zx
+    ? module.zx_quantize(data.data, data.width, data.height, opts.dither)
+    : module.quantize(data.data, data.width, data.height, opts.maxNumColors, opts.dither);
+
+  module.free_result();
+  module.doLeakCheck();
+
+  return new ImageData(
+    new Uint8ClampedArray(result.view),
+    result.width,
+    result.height
+  );
+}
+```
+
+This gives us a report like the following in the console:
+
+![Screenshot of a message "Direct leak of 12 bytes" coming from an anonymous wasm-function[…]](leak1-nodbg.png)
+
+Uh-oh, there are some small leaks, but the stacktrace is not very helpful as all the function names
+are mangled. Let's recompile with a basic debugging info to preserve them:
+
+```shell/12/
+emcc \
+  --bind \
+  ${OPTIMIZE} \
+  --closure 1 \
+  -s ALLOW_MEMORY_GROWTH=1 \
+  -s MODULARIZE=1 \
+  -s 'EXPORT_NAME="imagequant"' \
+  -I node_modules/libimagequant \
+  -o ./imagequant.js \
+  --std=c++11 \
+  imagequant.cpp \
+  -fsanitize=address \
+  -g2 \
+  node_modules/libimagequant/libimagequant.a
+```
+
+This looks much better:
+
+![Screenshot of a message "Direct leak of 12 bytes" coming from a `GenericBindingType<RawImage>::toWireType` function](leak1-dbg.png)
```

Comment:
Please use this formatting: https://web.dev/handbook/markup-media/#partial-screenshots